# Data Frame

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

In [1]:
import numpy as np
import pandas as pd

In [11]:
dict_area = {'jakarta' : 750,
                'bogor' : 490,
                'depok' : 350,
                'tangerang' : 270,
                'bekasi' : 670}

In [12]:
area =  pd.Series(dict_area)

In [13]:
area

jakarta      750
bogor        490
depok        350
tangerang    270
bekasi       670
dtype: int64

In [15]:
dict_population= {
    'jakarta' : 10,
     'bogor' : 7,
     'depok' : 8,
     'tangerang' : 9,
     'bekasi' : 8}

population=pd.Series(dict_population)

In [16]:
population

jakarta      10
bogor         7
depok         8
tangerang     9
bekasi        8
dtype: int64

In [29]:
region=pd.DataFrame({'area population':population, 'area':area})

In [30]:
region

area population  area
jakarta                 10   750
bogor                    7   490
depok                    8   350
tangerang                9   270
bekasi                   8   670

In [19]:
region['area']

jakarta      750
bogor        490
depok        350
tangerang    270
bekasi       670
Name: area, dtype: int64

In [20]:
region['area']['jakarta']

750

In [22]:
region['area population']['jakarta':'depok']
#explisit index

jakarta    10
bogor       7
depok       8
Name: area population, dtype: int64

In [23]:
region['area population'][0:3]
#implisit index

jakarta    10
bogor       7
depok       8
Name: area population, dtype: int64

In [31]:
region['area population'][0:3:2]
#slicing 3 parameter

jakarta    10
depok       8
Name: area population, dtype: int64

### Loc Property
- Pandas use the `loc` attribute to return one or more specified row(s)
- Syntax : dataframe.loc[row, column)

In [65]:
region.loc['jakarta','area']
#return row jakarta index and area column

750

### Iloc Property
- The iloc property gets, or sets, the value(s) of the specified indexes.

- Specify both row and column with an index.

In [60]:
region.iloc[0,1]
#return row 1 and column 2 

750

#### ------------------------  Task 1   -----------------------------

In [33]:
dict_regional = {'ana' : 'palembang',
                'akil' : 'bogor',
                'nia' : 'jakarta',
                'angga' : 'semarang',
                'una' : 'surabaya'}

In [34]:
dict_regional

{'ana': 'palembang',
 'akil': 'bogor',
 'nia': 'jakarta',
 'angga': 'semarang',
 'una': 'surabaya'}

In [35]:
regional = pd.Series(dict_regional)

In [36]:
dict_gender = {'ana' : 'F',
                'akil' : 'M',
                'nia' : 'F',
                'angga' : 'M',
                'una' : 'F'}

In [37]:
dict_gender

{'ana': 'F', 'akil': 'M', 'nia': 'F', 'angga': 'M', 'una': 'F'}

In [38]:
gender = pd.Series(dict_gender)

In [39]:
dict_num = {'ana' : '001',
                'akil' : '002',
                'nia' : '003',
                'angga' : '004',
                'una' : '005'}

In [40]:
dict_num

{'ana': '001', 'akil': '002', 'nia': '003', 'angga': '004', 'una': '005'}

In [45]:
nim = pd.Series(dict_num)

In [50]:
biodata = pd.DataFrame({'NIM':nim, 'Regional': regional, 'Sex' : gender, })

In [51]:
biodata

NIM   Regional Sex
ana    001  palembang   F
akil   002      bogor   M
nia    003    jakarta   F
angga  004   semarang   M
una    005   surabaya   F

In [56]:
biodata.iloc[0]
#return row 1

NIM               001
Regional    palembang
Sex                 F
Name: ana, dtype: object

In [53]:
biodata.loc['ana']
#return values of 'Ana' index

NIM               001
Regional    palembang
Sex                 F
Name: ana, dtype: object

In [58]:
biodata.iloc[1,0]
#return row 

'002'

In [55]:
biodata.loc['akil']

NIM           002
Regional    bogor
Sex             M
Name: akil, dtype: object

## Load data CSV

In [68]:
#load titanic data that has been aplouded in the same folder

df=pd.read_csv('Titanic.csv')

### head()
- viewing from top data
- can be customized
- `head` by default is top 5

In [70]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### tail()
- `tail()`returns a specified number of last rows.
- `tail()`returns the last 5 rows if a number is not specified.

In [72]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

## info()
- `info()` method prints information about the DataFrame.
- The information contains the number of columns, column labels, column data types, memory usage, range index, and the number of cells in each column (non-null values).

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


#### ---------------   Task 1     -------------------------

In [95]:
#show top 10 data
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [96]:
#Bottom 10 data
df.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

In [98]:
#which columns have missing values?
#and
#which columns have non-matching data types?
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


###### columns that contain missing values, namely: means count data != number of entry data
- Age : 714
- Cabin : 204
- Embarked : 889

###### column whose data type does not match: 
- Age should be int
- Cabin contains NaN

#### -----------------------------------------------------------------

### shape
`shape` is the number of rows and columns of the DataFrame

In [75]:
df.shape


(891, 12)

891 is the number of rows, 12 is the number of columns

### columns
`columns` returns the label of each column in the DataFrame.

In [78]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### index
- The `index` returns the index information of the DataFrame.
- The `index` information contains the labels of the rows. If the rows has NOT named indexes, the index property returns a RangeIndex object with the start, stop, and step values.

In [79]:
df.index

RangeIndex(start=0, stop=891, step=1)

### sum()
- Returns the `sum` of the values in the specified axis
- The `sum()` method adds all values in each column and returns the `sum` for each column.
- By specifying the column axis (axis='columns'), the `sum()` method searches column-wise and returns the `sum`| of each row.


In [81]:
df.sum()

C:\Users\windows\AppData\Local\Temp\ipykernel_12212\1703867807.py:1: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.sum()


PassengerId                                               397386
Survived                                                     342
Pclass                                                      2057
Name           Braund, Mr. Owen HarrisCumings, Mrs. John Brad...
Sex            malefemalefemalefemalemalemalemalemalefemalefe...
Age                                                     21205.17
SibSp                                                        466
Parch                                                        340
Ticket         A/5 21171PC 17599STON/O2. 31012821138033734503...
Fare                                                  28693.9493
dtype: object

In [82]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### isnull()
Finds NULL values

In [83]:
df.isnull().sum()
#number of null

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### notnull()
Finds values that are not NULL


In [85]:
df.notnull().sum()
#number of not null values

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

### describe()
Returns a description summary for each column in the DataFrame

In [87]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

#### ----------------------------------------    Task 2   ------------------------------------

#### Find insight information

#### By SibSp
- Passengers who are siblings are the most there are 8 siblings registered to board the ship
#### Based on PClass
- Pclass (passenger class) (50%) Q2 is 3rd class
#### By Fare
- Ticket price (British pound) top 512,329200
#### Based on Survived
- Q3 (75%) people survived

#### ---------------------------------------------------------------------

### mean()
- Return the mean of the values in the specified axis
- `Mean` : The average value

In [88]:
#returns the mean of the fare column
df['Fare'].mean()

32.204207968574636

### median()

- `Median` : The mid point value
- Return the median of the values in the specified axis

In [90]:
#returns the median of the fare column
df['Fare'].median()

14.4542

### mode()

- `Mode` : The most common value
- Returns the mode of the values in the specified axis

In [91]:
#returns the mean of the fare column
df['Fare'].mode()

0    8.05
Name: Fare, dtype: float64

### min()
Returns the min of the values in the specified axis

In [92]:
#returns the min values of the Fare column
df['Fare'].min()

0.0

### max()
Return the max of the values in the specified axis

In [93]:
#returns the max values of the Fare column
df['Fare'].max()

512.3292